src/components/connect.js

```diff
@@ -17,9 +17,36 @@ function getDisplayName(WrappedComponent) {
 // Helps track hot reloading.
 let nextVersion = 0
 
-export default function connect(mapPropsToRequestsToProps, options = {}) {
+function connectFactory(options = {}) {
+  function connectImpl(map) {
+    return connect(map, options)
+  }
+
+  connectImpl.withFetch = function (fetch) {
+    invariant(
+      typeof fetch === 'function',
+      '`withFetch` must be passed a function. Instead received a %s.',
+      typeof fetch
+    )
+
+    return connectFactory(assign({}, options, { fetch: fetch }))
+  }
+
+  connectImpl.withRef = function (withRef = true) {
+    return connectFactory(assign({}, options, { withRef: withRef }))
+  }
+
+  return connectImpl
+}
+
+export default connectFactory()
+
+function connect(mapPropsToRequestsToProps, options) {
   const finalMapPropsToRequestsToProps = mapPropsToRequestsToProps || defaultMapPropsToRequestsToProps
-  const { withRef = false } = options
+  const {
+    fetch = window.fetch,
```

Comment:
Ah, hmm, I'd missed the call to `window.Request` directly. If a completely custom `fetch` is provided, it would be reasonable to expect a similarly completely custom Request. Annoyingly, there's no way to derive `Request` from `fetch`. We could have a separate `.withRequest(Request)` to not use the global Request, or add it as optional parameter to `withFetch`.

The `fetch` function signature can be any of:

```js
fetch(req: Request)
fetch(url: String)
fetch(url: String, options: PlainObject) // same args passed to Request constructor
fetch(req: Request, options: PlainObject) // options here override Request ones 
```

So one can easily imagine switching between those forms to be done upon `firstArg instanceof Request`, where `Request` is _internal to the fetch implementation_. That's how [github/fetch aka whatwg-fetch](https://github.com/github/fetch/blob/master/fetch.js#L330) does it, for example. Also [this other fetch polyfill](https://github.com/camsong/fetch-ie8/blob/master/fetch.js#L330). Heck, I just tried replacing `window.Request` in my browser (which has a native `fetch`), and from the behaviour, it seems it does the same thing.

Thus, using `window.Request` with a non-`window` fetch is probably a no-go.